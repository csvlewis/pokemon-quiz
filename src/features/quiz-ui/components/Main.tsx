import { Paper, Stack, Typography } from "@mui/material";
import { useQuizContext } from "@/features/context/components/QuizContextProvider";
import { Options } from "@/features/quiz-ui/components/Options";
import { Submit } from "@/features/quiz-ui/components/Submit";

export function Main(): React.ReactNode {
  const { currentQuestion, questions, quizInProgress } = useQuizContext();

  if (quizInProgress)
    return (
      <Paper>
        <Stack
          justifyContent="center"
          alignItems="center"
          spacing={2}
          sx={{ margin: 2 }}
        >
          <Typography>
            Question {currentQuestion}/{questions.length}
          </Typography>
          <Typography>{questions[currentQuestion - 1]?.question}</Typography>
          <Options />
          <Submit />
        </Stack>
      </Paper>
    );
}
