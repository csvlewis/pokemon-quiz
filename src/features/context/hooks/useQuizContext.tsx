import { useContext } from "react";
import { QuizContextType } from "@/features/context/types/QuizContextType";
import { QuizContext } from "@/features/context/components/QuizContext";

export function useQuizContext(): QuizContextType {
  return useContext<QuizContextType>(QuizContext);
}
