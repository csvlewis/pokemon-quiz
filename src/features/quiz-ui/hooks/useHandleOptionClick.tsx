import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({});
    } else {
      setSelectedAnswer(option);
    }
  };
}
