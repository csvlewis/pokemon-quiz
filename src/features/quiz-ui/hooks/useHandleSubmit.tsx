import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useIncrementQuestion } from "@/features/quiz-ui/hooks/useIncrementQuestion";

export function useHandleSubmit(): () => void {
  const {
    correctAnswers,
    displayAnswer,
    selectedAnswer,
    setCorrectAnswers,
    setDisplayAnswer,
  } = useQuizContext();
  const incrementQuestion = useIncrementQuestion();

  return () => {
    if (displayAnswer) {
      incrementQuestion();
    } else {
      if (selectedAnswer.isCorrect) setCorrectAnswers(correctAnswers + 1);
      setDisplayAnswer(true);
    }
  };
}
