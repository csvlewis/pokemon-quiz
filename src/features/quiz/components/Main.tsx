import { Paper, Stack } from "@mui/material";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Question } from "@/features/quiz/components/Question";
import { Options } from "@/features/quiz/components/Options";
import { Submit } from "@/features/quiz/components/Submit";

export function Main(): React.ReactNode {
  const { quizInProgress } = useQuizContext();

  if (quizInProgress)
    return (
      <Paper>
        <Stack alignItems="center" spacing={2} sx={{ margin: "16px" }}>
          <Question />
          <Options />
          <Submit />
        </Stack>
      </Paper>
    );
}
