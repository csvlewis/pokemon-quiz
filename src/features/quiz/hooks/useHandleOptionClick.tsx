import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({ text: "", isCorrect: false });
    } else {
      setSelectedAnswer(option);
    }
  };
}
