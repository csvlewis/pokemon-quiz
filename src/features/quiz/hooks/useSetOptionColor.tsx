import { OverridableStringUnion } from "@mui/types";

import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetOptionColor(): (
  option: Option
) => OverridableStringUnion<"error" | "primary" | "success"> {
  const { displayAnswer } = useQuizContext();

  return (option: Option) => {
    let color: OverridableStringUnion<"error" | "primary" | "success"> =
      "primary";

    if (displayAnswer) {
      if (!option.isCorrect) {
        color = "error";
      } else {
        color = "success";
      }
    }

    return color;
  };
}
