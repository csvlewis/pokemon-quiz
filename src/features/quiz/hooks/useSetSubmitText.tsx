import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetSubmitText(): () => string {
  const { currentQuestion, displayAnswer, questions } = useQuizContext();

  return () => {
    let text: string = "Submit";

    if (displayAnswer) {
      if (currentQuestion >= questions.length) {
        text = "See Results";
      } else text = "Next Question";
    }

    return text;
  };
}
