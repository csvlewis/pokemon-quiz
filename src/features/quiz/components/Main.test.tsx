import React from "react";
import { Main } from "./Main";
import { render } from "@testing-library/react";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

jest.mock("@/features/context/hooks/useQuizContext");

const mockedUseQuizContext = useQuizContext as jest.Mock;

describe("Main", () => {
  it("should render correctly", () => {
    mockedUseQuizContext.mockReturnValue({
      quizInProgress: true,
      questions: [],
      selectedAnswer: { text: "", isCorrect: false },
    });

    const { container } = render(<Main />);
    expect(container).toMatchSnapshot();
  });
});
