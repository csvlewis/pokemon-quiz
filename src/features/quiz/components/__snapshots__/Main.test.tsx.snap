// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1gtchvp-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiStack-root css-d32kqm-MuiStack-root"
    >
      <div
        class="MuiStack-root css-6rg9iv-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
        >
          Question 
          /
          0
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
        />
        <img
          class="MuiBox-root css-1ri4hkh"
        />
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
        >
          Submit
        </p>
      </button>
    </div>
  </div>
</div>
`;
